src/arraystring.rs

```diff
@@ -22,7 +24,10 @@ use log::{debug, trace};
     derive(diesel::AsExpression, diesel::FromSqlRow)
 )]
 #[cfg_attr(feature = "diesel-traits", diesel(sql_type = diesel::sql_types::Text))]
-pub struct ArrayString<const N: usize> {
+pub struct ArrayString<const N: usize>
+where
+    Self: ValidCapacity,
```

Comment:
Sadly diesel's AsExpression derive does not take where into account, so this breaks the diesel-traits feature.

I've removed it as since all functions are behind this check anyway it's not the end of the world. But ideally AsExpression derive would propagate where to the derives like the FromSqlRow does, I'm thinking about providing a PR upstream but it will take time anyway before it's released.

So we can add this check here after it's been patched. I haven't taken a stab at it though, but it seems like a super easy fix

They need to do here: https://github.com/diesel-rs/diesel/blob/master/diesel_derives/src/as_expression.rs#L29

like they do here: https://github.com/diesel-rs/diesel/blob/master/diesel_derives/src/from_sql_row.rs#L37

Reply:
ah yes, i tried messing around with the clone function to see if uninit + partial copy is faster, but even for 255 it's a lot slower... just reverted that